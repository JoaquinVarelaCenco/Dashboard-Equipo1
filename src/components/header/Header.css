/* code css */
.header{
    position: fixed;
    width: 100%;
    height: 72px;
    display: flex;
    padding: 0 10px;
    opacity: .97;
    box-shadow: 0px 5px 5px -5px rgba(0, 0, 0, 0.75)

}

.headerGeneric{
    display: flex;
}

.headerGeneric button{
    background: none;
    border: none;
}

.title{
    align-self: center;
}

.title h2{
    font-weight:500;
}


/* HEADER PRODUCTS */
.headerProducts{
    display: flex;
    width: 100%;
    justify-content: flex-end;
}

.header__search-container{
    display: flex;
    align-items: center;
    position: relative;
    width: 280px;
  }

  .search-container__btnClose{
    display: none;
    height: 32px;
    width: 25px;
    align-self: center;
    border-radius: 50%;
    margin-right: 5px;
    text-decoration: none;
    font-family: 'Roboto', sans-serif;
    font-size: 16px;
    font-weight: 700;
    background-color: transparent;
    background-color: none;
    border: none;
    color: white;
  }
  
.header__search-container input{
    border: 1px solid rgba(0, 0, 0, 0.5);
    border-radius: 16px;
    height: 32px;
    width: 100%;
    padding: 0 0 0 20px;
}
 
.header__search-container input:placeholder{
    display: none;
}

.search-container__btnSearch{
    position: absolute;
    right: 8px;
    background: none;
    border: none;
}

.headerProducts-ContainerAgregar{
    align-self: center;
}

.headerProducts-ContainerAgregar a{
    text-decoration: none;
}
.headerProducts__btnAgregar{
    display: none;
    height: 32px;
    width: 150px;
    padding: 8px;
    align-self: center;
    border: 1px solid rgba(0, 0, 0, 0.5);
    border-radius: 16px;
    margin-left: 10px;
    text-decoration: none;
}

.headerProducts__btnAgregarAlternative{
    display: block;
    height: 32px;
    width: 30px;
    align-self: center;
    border: 1px solid rgba(0, 0, 0, 0.5);
    border-radius: 50%;
    margin-left: 10px;
    text-decoration: none;
    background-color: white;
    font-size: 25px;
    font-weight: 700;
}

/* HEADER EDIT PRODUCTS */
.headerEditProduct{
    display: flex;
    align-self: center;
}

.headerEditProduct__btnDeleteContainer{
    align-self: center;
    width: 100%;
    text-align: end;
}

.headerEditProduct__btnDelete{
    height: 32px;
    width: 100px;
    padding: 8px;
    
    border: 1px solid rgba(0, 0, 0, 0.5);
    border-radius: 16px;
    margin-left: 10px;
    
}

.headerEditProduct h2{
    font-weight: 500;
}
.header__menu-hamburguesa{
    margin-top: 22px;
    height: 32px;
    width: 32px;
    padding-right: 20px;
}


@media (min-width: 1025px){
    .header{
        width: calc(100% - 296px);
    }
    .headerProducts__btnAgregar{
        display: block;
    }
    .header__menu-hamburguesa{
        display: none;
    }

    .headerProducts__btnAgregarAlternative{
        display: none;
    }

    .header__search-container input:placeholder{
        display: block;
    }
}

@media (max-width: 500px){
    .header__search-container{
        width: 42px;
    }

}

/* Dark Mode */

.dark .headerEditProduct img,
.dark .header__menu-hamburguesa {
    filter: invert();
}

.dark .headerBtn{
    border: none;
    background-color: none !important;
}

.headerProducts .expandSearchBarStyle{
    width: 100%;
}