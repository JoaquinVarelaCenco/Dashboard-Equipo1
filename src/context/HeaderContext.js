import { useState, createContext } from "react";

export const HeaderContext = createContext("home");

export const HeaderProvider = ({ children }) => {
  const [page, setPage] = useState("home");

  const value = {
    page,
    currentPage: (p) => setPage(p),
  };

  return (
    <HeaderContext.Provider value={value}>{children}</HeaderContext.Provider>
  );
};
