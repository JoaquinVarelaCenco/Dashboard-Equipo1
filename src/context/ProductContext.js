import { useState, createContext } from "react";

export const ProductContext = createContext("home");

export const ProductProvider = ({ children }) => {
  const defaultValues = {
    title: "",
    price: 0,
    stock: 0,
    description: "",
    category: "",
    images: [],
    rating: {
      rate: 0,
      count: 0,
    },
  };

  const [product, setProduct] = useState(defaultValues);

  const [lastState, setLastState] = useState(defaultValues);

  const value = {
    product,
    currentProduct: (p) => setProduct(p),
    lastState,
    currentLastProduct: (i) => setLastState(i),
    resetCamps: () => setProduct(lastState),
    resetForm: () => {
      setProduct(defaultValues);
      setLastState(defaultValues);
    },
  };

  return (
    <ProductContext.Provider value={value}>{children}</ProductContext.Provider>
  );
};
