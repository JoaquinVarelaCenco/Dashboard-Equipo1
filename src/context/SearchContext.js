import { useState, createContext, useEffect } from "react";
import { getProducts } from "../services/apiServices";

export const SearchContext = createContext();

export const SearchProvider = ({ children }) => {
  const [products, setProducts] = useState([]);
  const [productsExist, setProductsExist] = useState(true);
  const [helpProducts, setHelpProducts] = useState([]);
  const [searchTermValue, setSearchTermValue] = useState("");
  const [orderBy, setOrderBy] = useState("Mas relevantes");

  const getAllProducts = () => {
    getProducts()
      .then((data) => {
        setProductsExist(true);
        orderProductsPrueba(data);
        setProducts(data);
        setHelpProducts(data);
      })
      .catch(() => {
        setProductsExist(false);
      });
  };

  const filterProducts = (searchTerm) => {
    const newProducts = helpProducts.filter((val) => {
      if (
        searchTerm === "" ||
        val.title.toLowerCase().includes(searchTerm.toLowerCase())
      ) {
        return val;
      } else if (
        val.description.toLowerCase().includes(searchTerm.toLowerCase())
      ) {
        return val;
      }
    });
    return newProducts;
  };

  const handleSearch = (e) => {
    const newFilter = filterProducts(e.target.value);
    setSearchTermValue(e.target.value);
    setProducts(newFilter);
  };

  const orderByHighPrice = (arrayData) => {
    arrayData.sort((a, b) =>
      a.price < b.price ? 1 : a.price > b.price ? -1 : 0
    );
  };

  const orderByLowPrice = (arrayData) => {
    arrayData.sort((a, b) =>
      a.price > b.price ? 1 : a.price < b.price ? -1 : 0
    );
  };

  const orderByMostRelevants = (arrayData) => {
    arrayData.sort((a, b) =>
      a.rating.count < b.rating.count
        ? 1
        : a.rating.count > b.rating.count
        ? -1
        : 0
    );
  };

  const orderByAlphabet = (arrayData) => {
    arrayData.sort((a, b) =>
      a.title > b.title ? 1 : a.title < b.title ? -1 : 0
    );
  };

  const orderProducts = (e) => {
    switch (e.target.value) {
      case "Mayor precio":
        setOrderBy("Mayor precio");
        orderByHighPrice(products);
        break;
      case "Menor precio":
        setOrderBy("Menor precio");
        orderByLowPrice(products);
        break;
      case "Mas relevantes":
        setOrderBy("Mas relevantes");
        orderByMostRelevants(products);
        break;
      case "A-Z":
        setOrderBy("A-Z");
        orderByAlphabet(products);
        break;
    }
  };

  const orderProductsPrueba = (data) => {
    
    switch (orderBy) {
      case "Mayor precio":
        setOrderBy("Mayor precio");
        orderByHighPrice(data);
        break;
      case "Menor precio":
        setOrderBy("Menor precio");
        orderByLowPrice(data);
        break;
      case "Mas relevantes":
        setOrderBy("Mas relevantes");
        orderByMostRelevants(data);
        break;
      case "A-Z":
        setOrderBy("A-Z");
        orderByAlphabet(data);
        break;
    }
  };



  useEffect(() => {
    getAllProducts();
  }, []);

  return (
    <SearchContext.Provider
      value={{
        handleSearch,
        products,
        productsExist,
        getAllProducts,
        helpProducts,
        searchTermValue,
        orderBy,
        orderProducts
      }}
    >
      {children}
    </SearchContext.Provider>
  );
};
