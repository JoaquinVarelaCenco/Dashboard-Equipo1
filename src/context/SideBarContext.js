import { useState, createContext } from 'react'

export const SideBarContext = createContext('unhide');

export const SideBarProvider = ({children}) => {

    const [ visibility, setVisibility ] = useState('unhide');

    const value = {
        visibility,
        toggleVisibility: () => setVisibility(prev => prev === 'unhide' ? 'hide' : 'unhide')
    }


return (
        <SideBarContext.Provider value={value}>
            {children}
        </SideBarContext.Provider>
    )
}