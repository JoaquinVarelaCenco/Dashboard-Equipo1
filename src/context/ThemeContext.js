import { useState, createContext } from 'react'

export const ThemeContext = createContext('clear');

export const ThemeProvider = ({children}) => {
    
    const initialMode = (localStorage.getItem("darkMode")) ? localStorage.getItem("darkMode") : 'clear';

    const [ theme, setTheme ] = useState(initialMode);

    const value = {
        theme,
        toggleTheme: () => setTheme(prev => prev === 'dark' ? 'clear' : 'dark')
    }
    localStorage.setItem("darkMode", theme);


return (
        <ThemeContext.Provider value={value}>
            {children}
        </ThemeContext.Provider>
    )
}