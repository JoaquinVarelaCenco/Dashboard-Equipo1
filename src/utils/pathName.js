export const pathName = (page) =>{
    //Lógica para establecer el pathname en que nos encontramos
    let pageTitle, titleLink;

    if (page === "/products") {
        pageTitle = "Productos";
        titleLink =page;
    } else if (page.includes("/stores")) {
        pageTitle = "Tiendas";
        titleLink =page;
    } else if (page === "/products/new") {
        pageTitle = "Productos";
        titleLink ="/products";
    } else if (page.includes("/products/")) {
        pageTitle = "Productos";
        titleLink ="/products";
    }else{
        pageTitle = "¡Hola Olivia!";
        titleLink = "/home";
    }
    return [pageTitle, titleLink];

}
