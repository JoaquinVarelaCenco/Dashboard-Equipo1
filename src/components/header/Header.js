import { SearchContext } from "../../context/SearchContext";
import "./Header.css";
import { useContext, useEffect, useRef } from "react";
import { HeaderContext } from "../../context/HeaderContext";
import { useLocation, Link } from "react-router-dom";
import searchImage from "../../assets/images/magnify.svg";
import menuImage from "../../assets/images/menu.svg";
import { SideBarContext, SideBarProvider } from "../../context/SideBarContext";
import { ThemeContext } from "../../context/ThemeContext";
import leftArrow from "../../assets/images/chevron-right (1).svg";

const Header = () => {
  let buttonMenu = useRef();
  // const btnSearch = useRef(null);

  const context = useContext(SearchContext);
  const { theme } = useContext(ThemeContext);
  const { page, currentPage } = useContext(HeaderContext);
  // const [productId, setProductId] = useState("");
  const location = useLocation();

  const pageTitle2 = "Products";
  const { toggleVisibility } = useContext(SideBarContext);
  const actualPage = page;
  let pageTitle = "¡Hola Olivia!";
  let productId = "";

  //USe effect para setear la url en la que me encuentro
  useEffect(() => {
    currentPage(location.pathname);
    titleContainer.current.style.display="flex";
  }, [location]);

 
 


  if (page === "/products") {
    pageTitle = "Products";
  } else if (page.includes("/stores")) {
    pageTitle = "Tiendas";
  } else if (page === "/products/new") {
    pageTitle = "Productos";
  } else if (page.includes("/products/")) {
    pageTitle = "Productos";
    productId = "#" + location.pathname.split("/")[2];
  }

  //Logica expandir input de búsqueda
  const titleContainer = useRef("");
  const inputSearch = useRef("");
  const inputSearchContainer = useRef("");
  const btnClose = useRef("");
  const containerAddProduct = useRef("");

  const expandSearchInput = () => {
    let width = window.screen.width;
    if (width <= 500) {
      titleContainer.current.style.display = "none";
      inputSearch.current.style.width = "100%";
      inputSearchContainer.current.style.width = "100%";
      btnClose.current.style.display = "block";
      containerAddProduct.current.style.display = "none"
    } else {
      titleContainer.current.style.display = "flex";
    }
  };

  const closeSearchInput = () => {
    titleContainer.current.style.display = "flex";
    inputSearch.current.style.width = "45px";
    inputSearchContainer.current.style.width = "45px";
    btnClose.current.style.display = "none";
    containerAddProduct.current.style.display = "block"
  };

   //useEffect para saber si screen width es mayor a 500px
   window.onresize = ()=>{
    if(window.screen.width> 500){
      closeSearchInput();
      inputSearch.current.style.width = "280px";
      inputSearchContainer.current.style.width = "280px";
    }else{
      closeSearchInput();
    }
  }
  return (
    <div className={`header ${theme}`}>
      <div className="headerGeneric" ref={titleContainer}>
        <button
          ref={buttonMenu}
          onClick={() => {
            toggleVisibility();
          }}
          className="header__menu-hamburguesa"
        >
          <img
            src={menuImage}
            alt="Menú hamburguesa"
            className="hamburguerMenu"
          />
        </button>
        <div className="title">
          <h2>{pageTitle}</h2>
        </div>
      </div>
      {page === "/products" ? (
        <div className="headerProducts">
          <div className="header__search-container" ref={inputSearchContainer}>
            <button
              onClick={closeSearchInput}
              className="search-container__btnClose"
              ref={btnClose}
            >
              X
            </button>
            <input
              type="text"
              className="header__search"
              placeholder="Buscar productos..."
              ref={inputSearch}
              onChange={context.handleSearch}
            />
            <button
              onClick={expandSearchInput}
              className="search-container__btnSearch"
            >
              <img src={searchImage} alt="Lupa de busqueda" />
            </button>
          </div>
          <div className="headerProducts-ContainerAgregar" >
            <Link to={"/products/new"}>
              <button className="headerProducts__btnAgregar">
                Agregar Producto
              </button>
            </Link>
          </div>
          <div className="headerProducts-ContainerAgregar" >
            <Link to={"/products/new"} ref={containerAddProduct}>
              <button className="headerProducts__btnAgregarAlternative">
                +
              </button>
            </Link>
          </div>
        </div>
      ) : page === "/products/new" ? (
        <div className="headerEditProduct">
          <img src={leftArrow} alt="" />
          <h2>Nuevo Producto</h2>
        </div>
      ) : page.includes("/products/") ? (
        <>
          <div className="headerEditProduct">
            <img src={leftArrow} alt="" />
            <h2>{productId}</h2>
          </div>
          <button className="headerEditProduct__btnDelete">ELIMINAR</button>
        </>
      ) : (
        ""
      )}
    </div>
  );
};

export default Header;
