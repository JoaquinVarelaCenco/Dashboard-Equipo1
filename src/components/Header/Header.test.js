import { render, screen, act } from "@testing-library/react";
import Header from "./Header";
import { BrowserRouter, MemoryRouter } from "react-router-dom";
import { pathName } from "../../utils/pathName";
import { HeaderProvider } from "../../context/HeaderContext";



describe("Test formulario de productos", ()=>{

    beforeEach(()=>{
        render(
        <MemoryRouter>
            <Header />
        </MemoryRouter>
            , { wrapper: HeaderProvider })
    })

    test("Test de prueba", ()=>{

    })

    //mostrar menú hamburguesa


    //funcionalidad menú hamburguesa

})