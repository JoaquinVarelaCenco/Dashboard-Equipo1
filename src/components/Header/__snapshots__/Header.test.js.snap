// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test formulario de productos El componente debe renderizarse correctamente 1`] = `
<div>
  <div
    class="header undefined"
  >
    <div
      class="headerGeneric displayFlex"
    >
      <button
        class="header__hamburguer-menu-btn noHoverEffect"
      >
        <img
          alt="Menú hamburguesa"
          class="header__hamburguer-menu-img"
          src="menu.svg"
        />
      </button>
      <div
        class="title"
      >
        <a
          href="/home"
        >
          <h2>
            ¡Hola Olivia!
          </h2>
        </a>
      </div>
    </div>
  </div>
</div>
`;
